### ISSUE TITLE
Question: why does syntax highlighting not work right after loading a new file?

### ISSUE THREAD

**Issue by sirop** [FXMisc/RichTextFX#1044]
Hallo.

I pretty much use https://github.com/FXMisc/RichTextFX/blob/master/richtextfx-demos/src/main/java/org/fxmisc/richtext/demo/JavaKeywordsAsyncDemo.java as a working example,
just added a function for loading another/new file:
````java
 private void readNIOFile(File file) {
    	try { 
    		  byte [] fileBytes = Files.readAllBytes(file.toPath());
    		 
    	          codeArea.replaceText(0,0, new  String(fileBytes));
    	    //  codeArea.setStyleSpans(0, computeHighlighting(codeArea.getText()));
          } catch (IOException e) {
			e.printStackTrace();
        }     
  }
````
<hr>

## Expected Behavior

Syntax highlighting should work immediately after a new file has been loaded.

## Actual Behavior

Syntax highlighting does not start immediately after a new file has been loaded, but only after making some (minimal) change to the text of CodeArea.

## Reproducible Demo

Provide a demo that maintainers of this project can copy, paste, and run to reproduce it immediately.

Use the following template to get started.
````java
import java.io.File;
import java.io.InputStream;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.io.IOException;
import java.nio.file.Files;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.nio.ByteBuffer;
import java.time.Duration;
import java.util.Collection;
import java.util.Collections;
import java.util.Optional;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javafx.application.Application;
import javafx.concurrent.Task;
import javafx.scene.Scene;

import javafx.scene.layout.Priority;
import javafx.scene.layout.VBox;

import javafx.stage.FileChooser;
import javafx.stage.Stage;

import javafx.geometry.Orientation;
import javafx.scene.control.Button;
import javafx.scene.control.Separator;

import javafx.scene.control.ToolBar;
import javafx.scene.control.Tooltip;

import org.fxmisc.flowless.VirtualizedScrollPane;
import org.fxmisc.richtext.CodeArea;
import org.fxmisc.richtext.LineNumberFactory;
import org.fxmisc.richtext.NavigationActions.SelectionPolicy;
import org.fxmisc.richtext.model.StyleSpans;
import org.fxmisc.richtext.model.StyleSpansBuilder;
import org.reactfx.Subscription;


import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;




public class  EditorDemo extends Application {  // JavaKeywordsAsyncDemo

    private static final String[] KEYWORDS = new String[] {
            "break", 
            "case",
            "declare", "default", "define", "do", "double", 
            "elif", "else", "endif", 
            "float", "for", "goto",
            "if", "if", "ifdef", "include", "int", "internal", 
            "return",
            "para_expr", "pragma", 
            "string", "switch", 
            "void", "while"
    };
   
    private static final String KEYWORD_PATTERN = "\\b(" + String.join("|", KEYWORDS) + ")\\b";
    private static final String PAREN_PATTERN = "\\(|\\)";
    private static final String BRACE_PATTERN = "\\{|\\}";
    private static final String BRACKET_PATTERN = "\\[|\\]";
    private static final String STRING_PATTERN = "\"([^\"\\\\]|\\\\.)*\"";
    private static final String COMMENT_PATTERN = "/\\*(.|\\R)*?\\*/" + "|" + "[\\;\\%][^\\n]*" + "|" + "//[^\\n]*"; // was ist mit %?
    private static final String GFUNC_PATTERN = "G\\d{2,4}|G\\d{3}(\\,?\\d)?";
    private static final String MFUNC_PATTERN = "M\\d{1,3}";


    private static final Pattern PATTERN = Pattern.compile(
            "(?<KEYWORD>" + KEYWORD_PATTERN + ")"
       //     +"(?<PREPRKEYWORD>" + PREPR_KEYWORD_PATTERN + ")"
            + "|(?<PAREN>" + PAREN_PATTERN + ")"
            + "|(?<BRACE>" + BRACE_PATTERN + ")"
            + "|(?<BRACKET>" + BRACKET_PATTERN + ")"
            + "|(?<STRING>" + STRING_PATTERN + ")"
            + "|(?<COMMENT>" + COMMENT_PATTERN + ")"
            + "|(?<GFUNC>" + GFUNC_PATTERN + ")"
            + "|(?<MFUNC>" + MFUNC_PATTERN + ")",
            Pattern.CASE_INSENSITIVE  // | Pattern.MULTILINE
    );

    private static final String sampleCode = String.join("\n", new String[] {
    		"#para_expr",
    		"#define PUNCH_CONF(slt, sld, sdt) G1005 A=slt D=sld O=sdt",
    		"#define PUNCH_CONF_LEAD(slt, sld) PUNCH_CONF(slt, sld, 0.0)",
    		"#define PUNCH_CONF_DELAY(sdt) PUNCH_CONF(0.0, 0.0, sdt)",
    		"#define PUNCH_ON() G60 G1005 H1",
    		"#define NIBBLE_ON() G60 G1005 H2",
    		"#define PUNCH_OFF() G61 G1005 H0",
    		"G00 X0 Y0 ;Fahren auf Ausgangspunkt",
    		"PUNCH_CONF_LEAD(5, 10) ;Parametrierung für das Setzen des Signals zur Stanzeinheit",
    		"PUNCH_ON() ;Start der Stanzfunktionalität",
    		"G01 G91 F2500 ;Parametrierung der Bewegung",
    		"X10 ;exakter Halt bei X10, Y0 plus Stanzvorgang",
    		"X10 ;exakter Halt bei X20, Y0 plus Stanzvorgang",
    		"X10 ;exakter Halt bei X30, Y0 plus Stanzvorgang",
    		"X10 ;exakter Halt bei X40, Y0 plus Stanzvorgang",
    		"Y10 ;exakter Halt bei X40, Y10 plus Stanzvorgang",
    		"X-10 ;exakter Halt bei X30, Y10 plus Stanzvorgang",
    		"X-10 ;exakter Halt bei X20, Y10 plus Stanzvorgang",
    		"X-10 ;exakter Halt bei X10, Y10 plus Stanzvorgang",
    		"X-10 ;exakter Halt bei X0, Y10 plus Stanzvorgang",
    		"PUNCH_OFF() ;Ausschalten der Stanzfunktionalität",
    		"M30 ;Programmende"
    });


    public static void main(String[] args) {
        launch(args);
    }

    private Stage mainStage;
    private CodeArea codeArea;
    private ExecutorService executor;
    private Subscription cleanupWhenDone;

    @Override
    public void start(Stage primaryStage) {
        executor = Executors.newSingleThreadExecutor();
        codeArea = new CodeArea();
        codeArea.setParagraphGraphicFactory(LineNumberFactory.get(codeArea));
     
        HighLightingOn();

        // call when no longer need it: `cleanupWhenFinished.unsubscribe();`
        
        Button loadBtn = createButton("loadfile", this::loadFile,
                "Load file.\n\n");
        Button saveBtn = createButton("savefile", this::saveFile,
                "Save file.\n\n");

        Button undoBtn = createButton("undo", codeArea::undo, "Undo");
        Button redoBtn = createButton("redo", codeArea::redo, "Redo");
        Button cutBtn = createButton("cut", codeArea::cut, "Cut");
        Button copyBtn = createButton("copy", codeArea::copy, "Copy");
        Button pasteBtn = createButton("paste", codeArea::paste, "Paste");
        
        ToolBar toolBar1 = new ToolBar(
                loadBtn, saveBtn,  new Separator(Orientation.VERTICAL),
               // wrapToggle, new Separator(Orientation.VERTICAL),
                undoBtn, redoBtn, new Separator(Orientation.VERTICAL),
                cutBtn, copyBtn, pasteBtn, new Separator(Orientation.VERTICAL)
               // alignLeftBtn, alignCenterBtn, alignRightBtn, alignJustifyBtn, new Separator(Orientation.VERTICAL),
              //  increaseIndentBtn, decreaseIndentBtn, new Separator(Orientation.VERTICAL),
               // paragraphBackgroundPicker
                );

        codeArea.replaceText(0, 0, sampleCode);

        VirtualizedScrollPane<CodeArea> vsPane  = new VirtualizedScrollPane<>(codeArea);
        VBox vbox = new VBox();
        VBox.setVgrow(vsPane, Priority.ALWAYS);
        vbox.getChildren().addAll(toolBar1, vsPane);
        Scene scene = new Scene(vbox, 600, 400);
        scene.getStylesheets().add(EditorDemo.class.getResource("java-keywords.css").toExternalForm());
        primaryStage.setScene(scene);
        primaryStage.setTitle("Editor Async Demo");
        primaryStage.show();
    }
    

    @Override
    public void stop() {
        executor.shutdown();
    }

    private Task<StyleSpans<Collection<String>>> computeHighlightingAsync() {
        String text = codeArea.getText();
        Task<StyleSpans<Collection<String>>> task = new Task<StyleSpans<Collection<String>>>() {
            @Override
            protected StyleSpans<Collection<String>> call() throws Exception {
                return computeHighlighting(text);
            }
        };
        executor.execute(task);
        return task;
    }

    private void applyHighlighting(StyleSpans<Collection<String>> highlighting) {
        codeArea.setStyleSpans(0, highlighting);
    }

    private static StyleSpans<Collection<String>> computeHighlighting(String text) {
        Matcher matcher = PATTERN.matcher(text);
        int lastKwEnd = 0;
        StyleSpansBuilder<Collection<String>> spansBuilder
                = new StyleSpansBuilder<>();
        while(matcher.find()) {
            String styleClass =
                    matcher.group("KEYWORD") != null ? "keyword" :
                  //  matcher.group("PREPRKEYWORD") != null ? "preprkeyword" :
                    matcher.group("PAREN") != null ? "paren" :
                    matcher.group("BRACE") != null ? "brace" :
                    matcher.group("BRACKET") != null ? "bracket" :
                    matcher.group("STRING") != null ? "string" :
                    matcher.group("COMMENT") != null ? "comment" :
                    matcher.group("GFUNC") != null ? "gfunc" :
                    matcher.group("MFUNC") != null ? "mfunc" :
                    null; /* never happens */ assert styleClass != null;
            spansBuilder.add(Collections.emptyList(), matcher.start() - lastKwEnd);
            spansBuilder.add(Collections.singleton(styleClass), matcher.end() - matcher.start());
            lastKwEnd = matcher.end();
        }
        spansBuilder.add(Collections.emptyList(), text.length() - lastKwEnd);
        return spansBuilder.create();
    }
    
    private void HighLightingOn() {
    	 cleanupWhenDone = codeArea.multiPlainChanges()
                 .successionEnds(Duration.ofMillis(500))
                 .supplyTask(this::computeHighlightingAsync)
                 .awaitLatest(codeArea.multiPlainChanges())
                 .filterMap(t -> {
                     if(t.isSuccess()) {
                         return Optional.of(t.get());
                     } else {
                         t.getFailure().printStackTrace();
                         return Optional.empty();
                     }
                 })
                 .subscribe(this::applyHighlighting);
    }
    
    private void HighLightingOff() {
    	cleanupWhenDone.unsubscribe();
    }
    
    private Button createButton(String styleClass, Runnable action, String toolTip) {
        Button button = new Button();
        button.getStyleClass().add(styleClass);
        button.setOnAction(evt -> {
            action.run();
            codeArea.requestFocus();
        });
        button.setPrefWidth(25);
        button.setPrefHeight(25);
        if (toolTip != null) {
            button.setTooltip(new Tooltip(toolTip));
        }
        return button;
    }
    
    private void loadFile() {
        String initialDir = System.getProperty("user.dir");
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Load source file");
        fileChooser.setInitialDirectory(new File(initialDir));
       // fileChooser.setSelectedExtensionFilter(
       //         new FileChooser.ExtensionFilter("Arbitrary RTFX file", "*" + RTFX_FILE_EXTENSION));
        File selectedFile = fileChooser.showOpenDialog(mainStage);
        if (selectedFile != null) {
           // codeArea.clear();
          //  HighLightingOff();
           // readExecutorNIOFile(selectedFile);
        	readNIOFile(selectedFile);
        }
    }
    
    private void readNIOFile(File file) {
    	try { 
    		  byte [] fileBytes = Files.readAllBytes(file.toPath());
    		  codeArea.clear();
    		  

    	      codeArea.replaceText(0,0, new  String(fileBytes));
    	    //  codeArea.setStyleSpans(0, computeHighlighting(codeArea.getText()));
    	     
    	      // HighLightingOn();
      } catch (IOException e) {
			e.printStackTrace();
		}
        
    }
    
    
    private void saveFile() {
        String initialDir = System.getProperty("user.dir");
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Save source file");
        fileChooser.setInitialDirectory(new File(initialDir));
       // fileChooser.setInitialFileName("example rtfx file" + RTFX_FILE_EXTENSION);
        File selectedFile = fileChooser.showSaveDialog(mainStage);
        if (selectedFile != null) {
            writeFile(selectedFile);
        }
    }   
    
    private void writeFile(File file) {
    //	List<String> contentsList = Arrays.asList(codeArea.getText().split(System.lineSeparator()));
    	int size = codeArea.getVisibleParagraphs().size();
    	String lsep = System.lineSeparator();
    	
    	try( FileWriter writer = new FileWriter(file.getName());
    		 BufferedWriter bufferedWriter = new BufferedWriter(writer)) {
    	     for ( int i = 0; i < size ; i++ ) {
    		      String parag = codeArea.getText(i) + lsep;
    		      bufferedWriter.write(parag   , 0, parag.length() - 1);
    	     }
    	 } catch (IOException e) {
 			e.printStackTrace();
 		}
    		
     }
        

    

}



````
Resources used are those of RichText Demo .
## Environment info:

- RichTextFX Version: a89a333e47441af891134d0cacac8d8f9b740e20
- Operating System: WIN 10 64 bit
- Java version:
````
 $ javac -version
javac 1.8.0_302 
````
